### PR TITLE
Carbg 49 implementation of sqlite code as a storage method

### PR DESCRIPTION
**Pull request by NicoMeleThinkAndDev**
Pull request to add sqlite as a possible storage method for boardgame

- Added sqlite3.ts which implements all the methods necessary for the correct operation of the boardgame.
- Added sqlite3.test.ts to test that all functionality works as expected.
- Also modified the server.ts file and the index.ts file to support Sqlite so that it can be used in the back-end App.ts file.

### CODE REVIEW

**Review by jm-aragonpaz — APPROVED** [review]
Awesome Work

File: package.json [hunk 1]
```diff
@@ -32,7 +32,8 @@
     "postpack": "npm run clean",
     "prettier": "prettier --write \"{examples,src,benchmark}/**/*.{ts,tsx,js,jsx,css,md}\"",
     "changelog": "node ./scripts/changelog.js",
-    "clean": "node ./scripts/clean.js"
+    "clean": "node ./scripts/clean.js",
+    "publish:playground":"npm pack && cp *.tgz ../cartesi-boardgame-playground/backend"
```
Comment (jm-aragonpaz) [inline package.json]:
```suggestion

```

File: package.json [hunk 2]
```diff
@@ -32,7 +32,8 @@
     "postpack": "npm run clean",
     "prettier": "prettier --write \"{examples,src,benchmark}/**/*.{ts,tsx,js,jsx,css,md}\"",
     "changelog": "node ./scripts/changelog.js",
-    "clean": "node ./scripts/clean.js"
+    "clean": "node ./scripts/clean.js",
```
Comment (jm-aragonpaz) [inline package.json]:
```suggestion
    "clean": "node ./scripts/clean.js"
```

**Review by jm-aragonpaz — APPROVED** [review]

**Review by jm-aragonpaz — APPROVED** [review]
LGTM

File: src/master/master.ts [hunk 3]
```diff
@@ -303,13 +299,18 @@ export class Master {
...
+        const writes = [
+          this.storageAPI.setState(key, stateWithoutDeltalog, deltalog),
+        ];
+        if (newMetadata) {
+          writes.push(this.storageAPI.setMetadata(key, newMetadata));
+        }
+
+        await Promise.all(writes);
+      } catch (error) {
+        console.error('Error setting state:', error);
```
Comment (lucas-marc) [inline src/master/master.ts:312]: issue: we are stoping the propagation of the exception here, might make the consumer understand that the process finished correctly, when that might not be the case.

File: src/master/master.ts
Comment (fgregoribattista) [inline src/master/master.ts:1]: In the near future: I would like to make a refact to the structure you have.
First I see that it has interfaces, types, it is a class. Surely for a better human readability I would put the types and interfaces in separate files. Also there is code that can be refactored, like the auth validation to respond unauthorized, in some cases different messages respond to the same validation.

File: src/server/db/sqlite3.test.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+  beforeAll(async () => {
+    db = new Sqlite();
+    await db.connect();
+  });
+  afterEach(async () => {
+    await db.clear();
+  });
+
+  test('basic', async () => {
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:23]: give a more descriptive name to the test

Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:23]: there are a lot of checks, which could easily be separate tests or functions where different tests call them.

File: src/server/db/sqlite3.test.ts [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        unlisted: null,
+        nextMatchID: null,
+        createdAt: null,
+        updatedAt: null,
+      });
+    }
+
+    // // Must return all keys
+    const keys = await db.listMatches();
+    expect(keys).toEqual(['matchID']);
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:65]: which is 'matchId', I see it repeated in several values. 
We should have a constant where the name of the constant gives meaning to what it is. Beyond its value

File: src/server/db/sqlite3.test.ts [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+    await db.setState('matchID', null, [logEntry1]);
+    await db.setState('matchID', null, [logEntry2]);
+    await db.setState('matchID', null, [logEntry3]);
+
+    const result = await db.fetch('matchID', { log: true });
+    expect(result.log).toEqual([
+      logEntry1,
+      logEntry2,
+      {
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:128]: this obj is the same than one of the above? why don't we use it?

File: src/server/db/sqlite3.test.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    await db.wipe('matchID');
+
+    // Shall create match, then clear DB, then check whether DB is cleared
+    await db.setState('game2', state as State);
+    const keys2 = await db.listMatches();
+    expect(keys2).toHaveLength(0);
+  });
+
+  test('log', async () => {
+    const logEntry1: LogEntry = {
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:83]: I would consider having the constants outside the test, to make the test easier to read.

File: src/server/db/sqlite3.test.ts [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+      await db.createMatch('matchID3', {
+        initialState: state as State,
+        metadata: {
+          gameName: 'game2',
+          updatedAt: new Date(2020, 4).getTime(),
+        } as Server.MatchData,
+      });
+    });
+
+    test('filter by gameName', async () => {
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:173]: there are different tests mixed up in the same test, I would separate them.

File: src/server/db/sqlite3.test.ts [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        expect.arrayContaining(['matchID', 'matchID2', 'matchID3'])
+      );
+
+      keys = await db.listMatches({ gameName: 'game1' });
+      expect(keys).toEqual(expect.arrayContaining(['matchID', 'matchID2']));
+
+      keys = await db.listMatches({ gameName: 'game2' });
+      expect(keys).toEqual(['matchID3']);
+    });
+    test('filter by isGameover', async () => {
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:185]: same here

File: src/server/db/sqlite3.test.ts [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+      );
+
+      keys = await db.listMatches({ where: { isGameover: true } });
+      expect(keys).toEqual(['matchID2']);
+
+      keys = await db.listMatches({ where: { isGameover: false } });
+      expect(keys).toEqual(expect.arrayContaining(['matchID', 'matchID3']));
+    });
+    test('filter by updatedBefore', async () => {
+      const timestamp = new Date(2020, 4);
```
Comment (fgregoribattista) [inline src/server/db/sqlite3.test.ts:199]: same var, please put it outside and reuse it

File: src/master/master.ts (hunk 3 above)
Comment (NicoMeleThinkAndDev) [reply src/master/master.ts:312]: that's right, i fixed it

**Pull request merged**